### ISSUE TITLE
Claim: m.component is fragile, suggestion for fix

### ISSUE THREAD

**Issue by ralphbecket** [lhorie/mithril.js#757]
I've been experimenting with `m.component` and I think it's too fragile to use (this is a spin-off of a discussion started at https://groups.google.com/forum/#!topic/mithriljs/poLZ9u2yXwk ).

Here's my understanding of how `m.component(cmpnt, args)` works when `cmpnt` supplies a `controller` constructor:
- the first time the component instance is rendered in a view, `m.component` will call `cmpnt.controller` to create a controller instance and cache the result;
- on subsequent renderings, **provided** the component instance is in the same place in the vDOM *or* has the same vDOM parent and has been provided with a sibling-unique `key` attribute, then `m.component` will call `cmpnt.view` with the cached controller instance, **otherwise** `m.component` assumes this is a new instance of the component and will construct a new controller instance for it as above;
- this means the lifetime of a component controller instance is strongly dependent on its context in the vDOM.

I claim that this context dependence is a Very Bad Thing.

Here is an example http://jsfiddle.net/ralphbecket/y4qgh6e8/3/ where the context of a pair of components is randomly changed by a level of nesting.  Clicking on one of the buttons, you will observe its counter resetting at random as `m.component` behaves as just described.  This is clearly not good and I think the fundamental problem is that Mithril's rendering engine should *not* be responsible for managing component controller lifetimes.

I have an alternative approach here, http://jsfiddle.net/ralphbecket/y4qgh6e8/6/ .  The code is, if anything, shorter and clearer.  I claim it is morally equivalent to the `{ controller: ..., view: ... }` scheme, but with the strongly desirable context independence.

I'd go a step further than this: instead of having to return `{ view: function () { ... } }` objects, could Mithril not be extended to also recognise plain function objects in views?  That would remove another layer of unneccessary decoration.

**Comment by Bondifrench** [comment]
I am not sure the keys are for that kind of identity, I would think the keys would be if your components were in an array, it makes them unique and you can order your components properly.
Here the fact that you have 2 components `a` in your `return`, I am wondering if in effect you have two separate instances of your component.

On the second topic, I actually like to have `view: function () { return ...}`, in a way it makes it very similar to React with  `render: function () { return (...);}` and it's very explicit it's the view.

**Comment by ralphbecket** [comment]
Well, my primary complaint is the behaviour.  Regarding views, we are happy with prop's and promises being vanilla functions, so making views into vanilla functions might be seen as adding a pleasing consistency -- but that's a much smaller point.

Yes, in my examples I have two distinct instances of the same component, but I could just as easily have distinct instances of distinct components and the problem would manifest itself just the same.

**Comment by lhorie** [comment]
Something that I admit isn't clear in the current documentation is the pre-v0.2 way of using components, which is something like this: http://jsfiddle.net/y4qgh6e8/10/

The v0.2 component system is designed to make components behave similar to elements in terms of their lifecycle. The most important aspect that v0.2 handles is managing controller lifecycle to be in sync w/ the component's presence in the view (particularly to make removals from lists simple)

Currently there's no support for moving either elements or components to different parents. You can achieve that as you have done, or using the old component pattern from my link above and managing component lifecycle yourself (if that is needed in your application at all)

**Comment by ralphbecket** [comment]
@lhorie thanks for your reply.

Let me see if I have this right.  The current component scheme aims to align the state-losing behaviour of ordinary HTML elements with that of components.  For example, in this fiddle, http://jsfiddle.net/ralphbecket/d718p9u0/1/ , one checkbox changes its position in the DOM and consequently loses its state on redraws.  If I want that checkbox to retain its state across DOM context changes, I must track and bind it explicitly in my application.

Okay, I buy that.  I still feel uncomfortable about the semantics of `m.component`, though.  It feels too much like a somewhat flaky micro-optimisation valid only for the case where you can guarantee the DOM context won't change.  Indeed, I'm starting to wonder the same thing about the `key` scheme.  My instinct tells me it should be possible to have the `key` optimisation work with globally unique keys that allow for DOM context changes -- but I haven't worked out the details on paper, so I'm not a hundred per cent convinced yet.

**Comment by barneycarroll** [comment]
@ralphbecket I totally agree with your analysis of component state auto-management in Mithril, but I think it's important to consider the goals of view-based component management in Mithril.

To elaborate on what @lhorie was saying, before auto-management came along in v0.2, Mithril managed top level component lifecycle, but it was up to you to initialise the controllers for the rest. This was good because it forced you to understand and grapple with Mithril lifecycle, and bad... because it forced you to understand and grapple with Mithril lifecycle.

Here's how your original fiddle looks like without relying on v0.2 auto-management:

https://jsbin.com/mujere/edit?js,output

Now it's important to appreciate that, as verbose as this might be, this pattern is significantly more scalable than the closure-based method in your original example, because whereas your application instantiated the state for `app` and the `counterButton` components on initialisation, this initialises state as part of the Mithril lifecycle. This means you could have several `app` components on a page, and create new ones dynamically - component states are created as and when needed. This is a crucial necessity in complex apps: the ability to create and retrieve different configurations of component states on the fly.

But it's cumbersome! Mithril's lifecycle is awesome, but the most awesome part of it is that all the clever and difficult stuff happens under the hood, and when we code it just works. Writing ambitious applications using this API was too much like Backbone - the application logic is drowned in framework implementation API and you end up not being able to see the wood for the trees.

So v0.2 brought a huge degree of simplification for the most simple scenarios - Mithril can now 'guess' when you want controllers initialised, retrieved and associated, but only for lists of optionally-keyed components with a shared parent and a unique DOM element root. But I agree, it's way too simplistic and hacky and simplistic for many scenarios - and having to mix and match new and old styles (or refactor one to the other) is an ugly kind of pain.

Before v0.2 hit, I already badly needed to have flexible and terse component initialisation, so I wrote a plugin, [Modulator], that tries to provide the tersest invocation signature (in the view) while offering an API that covers any kind of controller management policy you could ask for. Here's how your code looks with Modulator:

https://jsbin.com/pofujo/edit?js,output

Modulator is a functor (ie a function which returns a function): the first set of parens consume all the arguments needed to identify a component instance, the second takes all the arguments you want to pass to the component's controller (when it initialises) and view.

Mithril says that the context for a list of components is a shared parent DOM node. Modulator allows you to specify whatever context you want - the recommended context being the parent component's `ctrl`. Like Mithril, Modulator assumes you want to iterate within the context if you don't pass a key. I believe you can establish any kind of conceivable identification using this method. There are also shorthands for globally unique components (so you can always retrieve the same context), `mod.unique( component )`, and global context component (so you can retrieve a component instance in a different context using the same key), `mod.global( component, key? )`.

I'd be interested in your thoughts.

**Comment by lhorie** [comment]
@ralphbecket In theory, yes, global keys should be doable. My only reservation is that global keys introduce the problem of global uniqueness (i.e. you can't have the same component repeat twice in a page if it uses global keys).

**Comment by ralphbecket** [comment]
Sorry about the late response, I've just returned from a week's hols.

@barneycarroll - I don't see why my closure-based approach is any less scalable than yours; I can easily support multiple `app` instances, add new ones dynamically, etc: http://jsfiddle.net/y4qgh6e8/11/

My key complaint, though, is that the lifetime of a component should be independent of the structure of its context.  It simply should not happen that components can be randomly reset because of changes in the surrounding context.  It would be better not to have this sort of convenience magic at all: if one does get bitten by this situation, diagnosis will be painful and the remedy will require substantial recoding.

**Comment by ralphbecket** [comment]
@lhorie - Do you mean you want to have the same component *instance* appear in multiple places on the same page?  Yes, I agree this is a problem, but probably less common in my estimation.

**Comment by StephanHoyer** [comment]
From a users point of view I found `m.component` pretty awesome. The problem with not being movable aligns with the behaviour of elements. Thats a good thing since I use components like web components. I also sometimes replace DOM elements with components if the behaviour gets more complex.

I only rarely encounter the problem you describe. Moving DOM elements around can be avoided in most cases.

As the other guys already say, the thing you want to achieve is doable with other, more complicated approaches.

I'm pretty happy with how `m.component` works.

**Comment by ralphbecket** [comment]
I have noticed that the web world is satisfied with "it mostly works, most of the time", but I get the impression Mithril's philosophy is based on a stricter degree of consistency and predictability.  It is quite possible that the design space contains a solution that allows for the convenience of leaving component management to `m.component` while preserving a robust, composable semantics.  It would be a shame to commit prematurely to a solution without the latter property.

**Comment by barneycarroll** [comment]
> I don't see why my closure-based approach is any less scalable than yours

@ralphbecket I think you sent an old fiddle — that link doesn't seem any different from the earlier ones. Let me rephrase: the ultimate convenience of automatic component lifecycle management is that in Mithril, you can write a particular piece of view code, and behind the scenes Mithril will intelligently either initialise controllers or retrieve them and render them all in one same expression. In your code, you are explicitly initialising A and B outside of component space before mounting their parent component.

**Comment by ralphbecket** [comment]
@barneycarroll -- Are you sure?  That http://jsfiddle.net/y4qgh6e8/11/ version has two `app` instances running simultaneously when I click on it.

I agree that my code is doing what you say it does, the reason for that is that Mithril is not sufficiently intelligent with regards to controller initialisation :-)  In fact, it can't be that intelligent: the API gives me no composable way of telling Mithril what I intend.

Consider this example: I write a handy component, a grid, say.  Someone else, Joe, uses it in his app and it all works as expected.  Hooray.  Then the boss assigns Mary to implement user-configurable UI layout.  Oh, no!  Now, when the user rearranges the UI (through whatever mechanism Mary provided) my grid component is reset.

I argue that Joe and Mary should be able to work independently of each other in this way and *not* be surprised like this.  Moreover, they shouldn't have to decide from the outset whether to trust `m.component` magic or not depending on whether the design spec may or may not evolve to include dynamic rearrangement of the UI.

**Comment by wambotron** [comment]
Sorry for dredging up an older topic, but I have a question regarding caching of views that seems tangentially related to this issue.

I have a scenario where we want to have a mobile view which has components arranged `A, B, C, D` and a desktop view which has components arranged `A,D,B,C`. I have a resize event which changes the state of the app (desktop vs mobile). The problem I'm running into is that the view is cached, and so the view function which determines the component position is never run after the initial page load.

Is there a good way to force it to reload that view without sacrificing too much performance?

[Here is a fiddle demonstrating what I mean](http://jsfiddle.net/swLjt778/). When you `run` it, it correctly shows the order. If you resize the HTML window without hitting `run`, it never changes the order.

**Comment by barneycarroll** [comment]
It's not that the view function is cached, it's that your application
mounts and then never redraws. Simply chuck in an m.redraw() call at the
end of the resize handler and it works: http://jsfiddle.net/barney/x342xb7n/

A bit of optimisation: only redraw if the value's changed - also, you don't
need to call m.mount (formerly m.component) if you're invoking the
component without arguments, and the array notation is superfluous:
http://jsfiddle.net/barney/t2j6o66a/

Regards,
Barney Carroll

barney.carroll@gmail.com
+44 7429 177278

barneycarroll.com

**Comment by wambotron** [comment]
*headdesk*I thought I tried that. That works, thanks!
